staging/src/k8s.io/sample-controller/pkg/signals/signal.go

```diff
@@ -26,18 +26,8 @@ var onlyOneSignalHandler = make(chan struct{})
 // SetupSignalHandler registered for SIGTERM and SIGINT. A stop channel is returned
 // which is closed on one of these signals. If a second signal is caught, the program
 // is terminated with exit code 1.
-func SetupSignalHandler() (stopCh <-chan struct{}) {
+func SetupSignalHandler() (context.Context, context.CancelFunc) {
```

Comment:
The "If a second signal is caught, the program is terminated with exit code 1" part no longer seems to be true. I looked up the documentation of `signal.NotifyContext` and function simply doesn't do that.

Changing from a stop channel to a context is useful, but I think you should just return a context and handle canceling it entirely in this function, using `signal.Notify` as it was done before.

Reply:
When I tested it, I found this logic redundant, specifically when we I send SIGINT the function closes the channel (stopCh) and exits the program.

The core of the work seems to happen in the _Controller.Run function_ where the workers threads are spawned. At this point closing the channel is as good as exiting, as we quickly return to main and exit the program. Of course, my test environment maybe skewing my perception, as there could be a use case where we are waiting for something before we reach this spot. Probably the maintainers or authors who have used it can comment if this is actually used or needed.

Unless someone can verify this is ok, I will update the function to have the same logic as before. Instead of stopCh I will use the ctx Done channel.

```
func (c *Controller) Run()... 
	// Launch two workers to process Foo resources
	...
	klog.Info("Started workers")
	<-stopCh
	klog.Info("Shutting down workers")
```

Reply:
This is an example of how to write a controller. A real controller might get stuck after the initial SIGINT and then being able to kill it with another one may be useful. This code shows how to implement that, even when the sample controller never runs into this. I think we should keep this feature.

Reply:
Yes, this makes sense. I have updated the code, please review.